Incorrect bracket matching and broken syntax highlighting with this chunk of C text

### Summary

Typing `mm` to go to matching bracket on the top bracket goes to the wrong one, and fiddeling with the file changing the grammar of things (e.g `transform_scope` => `transform_scope()`) fixes it. So does removing any line inside of the `pipeline_scope`.

When the highlighting on the outermost scope is colored blue as in the first screenshot:
![image](https://user-images.githubusercontent.com/28424206/207977047-997cbc5d-4aad-404f-91d0-17cecf69261b.png)

That means it's broken as seen in the incorrectly highlighted bracket, and when it's green that means it's working as shown below:
![Uploading image.png…]()


### Reproduction Steps

Open helix and paste this buffer of C code into the file, saving it as `tmp.c` or something so it recognizes it as C:

```c
{
  transform_scope
  {
    pipeline_scope(goodpixel_pipeline)
    {
      sgp_set_image(0, image_rightclick);
      cpVect draw_at = cpvadd(entity_pos(b), cpv(BOX_SIZE, 0));
      rotate_at(-entity_rotation(b) - rotangle(b->compass_rotation), draw_at.x, draw_at.y);
      draw_texture_centered(draw_at, BOX_SIZE + sin(exec_time * 5.0) * BOX_SIZE * 0.1);
      sgp_reset_image(0);
    }
  }
}

```

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2022-12-15T14:05:58.168 helix_lsp::transport [ERROR] err <- "I[14:05:58.168] --> reply:textDocument/signatureHelp(25) 18 ms\r\n"
2022-12-15T14:05:58.205 helix_lsp::transport [ERROR] err <- "I[14:05:58.205] ASTWorker building file C:\\Users\\Cameron\\Documents\\flight\\tmp.c version 75 with command inferred from C:/Users/Cameron/Documents/flight/thirdparty/enet/callbacks.c\r\n"
2022-12-15T14:05:58.205 helix_lsp::transport [ERROR] err <- "[C:/Users/Cameron/Documents/flight/]\r\n"
2022-12-15T14:05:58.205 helix_lsp::transport [ERROR] err <- "\"C:\\\\Users\\\\Cameron\\\\AppData\\\\Roaming\\\\ClangPowerTools\\\\LLVM_Lite\\\\Bin\\\\clang++.exe\" --driver-mode=g++ -Wall -fms-compatibility-version=19.10 -Wmicrosoft -Wno-invalid-token-paste -Wno-unknown-pragmas -Wno-unused-value -fsyntax-only -DUNICODE -D_UNICODE -D_MT -D_DEBUG -DDEBUG -D_CONSOLE -D_DEBUG_FUNCTIONAL_MACHINERY \"-isystemC:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.33.31629/include\" \"-isystemC:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.33.31629/atlmfc/include\" \"-isystemC:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/VS/include\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/ucrt\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/um\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/shared\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/winrt\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/cppwinrt\" -Ithirdparty/enet/include -Ithirdparty/Chipmunk2D/include/chipmunk -Ithirdparty -Ithirdparty/Chipmunk2D/include -Ithirdparty/minilzo -Ithirdparty/opus/include \"-resource-dir=C:\\\\Users\\\\Cameron\\\\Documents\\\\lib\\\\clang\\\\15.0.3\" -- \"C:\\\\Users\\\\Cameron\\\\Documents\\\\flight\\\\tmp.c\"\r\n"
2022-12-15T14:05:58.227 helix_lsp::transport [ERROR] err <- "I[14:05:58.227] --> textDocument/publishDiagnostics\r\n"
2022-12-15T14:06:51.146 helix_lsp::transport [ERROR] err <- "I[14:06:51.146] <-- textDocument/signatureHelp(26)\r\n"
2022-12-15T14:06:51.162 helix_lsp::transport [ERROR] err <- "I[14:06:51.162] --> reply:textDocument/signatureHelp(26) 16 ms\r\n"
2022-12-15T14:06:51.296 helix_lsp::transport [ERROR] err <- "I[14:06:51.296] <-- textDocument/didChange\r\n"
2022-12-15T14:06:51.298 helix_lsp::transport [ERROR] err <- "I[14:06:51.298] <-- textDocument/signatureHelp(27)\r\n"
2022-12-15T14:06:51.314 helix_lsp::transport [ERROR] err <- "I[14:06:51.314] --> reply:textDocument/signatureHelp(27) 15 ms\r\n"
2022-12-15T14:06:51.351 helix_lsp::transport [ERROR] err <- "I[14:06:51.351] ASTWorker building file C:\\Users\\Cameron\\Documents\\flight\\tmp.c version 76 with command inferred from C:/Users/Cameron/Documents/flight/thirdparty/enet/callbacks.c\r\n"
2022-12-15T14:06:51.351 helix_lsp::transport [ERROR] err <- "[C:/Users/Cameron/Documents/flight/]\r\n"
2022-12-15T14:06:51.351 helix_lsp::transport [ERROR] err <- "\"C:\\\\Users\\\\Cameron\\\\AppData\\\\Roaming\\\\ClangPowerTools\\\\LLVM_Lite\\\\Bin\\\\clang++.exe\" --driver-mode=g++ -Wall -fms-compatibility-version=19.10 -Wmicrosoft -Wno-invalid-token-paste -Wno-unknown-pragmas -Wno-unused-value -fsyntax-only -DUNICODE -D_UNICODE -D_MT -D_DEBUG -DDEBUG -D_CONSOLE -D_DEBUG_FUNCTIONAL_MACHINERY \"-isystemC:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.33.31629/include\" \"-isystemC:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.33.31629/atlmfc/include\" \"-isystemC:/Program Files/Microsoft Visual Studio/2022/Community/VC/Auxiliary/VS/include\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/ucrt\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/um\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/shared\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/winrt\" \"-isystemC:/Program Files (x86)/Windows Kits/10/Include/10.0.22000.0/cppwinrt\" -Ithirdparty/enet/include -Ithirdparty/Chipmunk2D/include/chipmunk -Ithirdparty -Ithirdparty/Chipmunk2D/include -Ithirdparty/minilzo -Ithirdparty/opus/include \"-resource-dir=C:\\\\Users\\\\Cameron\\\\Documents\\\\lib\\\\clang\\\\15.0.3\" -- \"C:\\\\Users\\\\Cameron\\\\Documents\\\\flight\\\\tmp.c\"\r\n"
2022-12-15T14:06:51.354 helix_lsp::transport [ERROR] err <- "I[14:06:51.354] <-- textDocument/signatureHelp(28)\r\n"

```

</details>


### Platform

Windows

### Terminal Emulator

Windows Terminal

### Helix Version

22.12-47-g4acdc03b

--- a/helix-core/src/match_brackets.rs
+++ b/helix-core/src/match_brackets.rs
@@ -1,8 +1,11 @@
+use std::iter;
+
 use tree_sitter::Node;
 
 use crate::{Rope, Syntax};
 
 const MAX_PLAINTEXT_SCAN: usize = 10000;
+const MATCH_LIMIT: usize = 16;
 
 // Limit matching pairs to only ( ) { } [ ] < > ' ' " "
 const PAIRS: &[(char, char)] = &[
@@ -50,26 +53,55 @@ fn find_pair(syntax: &Syntax, doc: &Rope, pos: usize, traverse_parents: bool) ->
     let tree = syntax.tree();
     let pos = doc.char_to_byte(pos);
 
-    let mut node = tree.root_node().named_descendant_for_byte_range(pos, pos)?;
+    let mut node = tree.root_node().descendant_for_byte_range(pos, pos)?;
 
     loop {
-        let (start_byte, end_byte) = surrounding_bytes(doc, &node)?;
-        let (start_char, end_char) = (doc.byte_to_char(start_byte), doc.byte_to_char(end_byte));
+        if node.is_named() {
+            let (start_byte, end_byte) = surrounding_bytes(doc, &node)?;
+            let (start_char, end_char) = (doc.byte_to_char(start_byte), doc.byte_to_char(end_byte));
 
-        if is_valid_pair(doc, start_char, end_char) {
-            if end_byte == pos {
-                return Some(start_char);
+            if is_valid_pair(doc, start_char, end_char) {
+                if end_byte == pos {
+                    return Some(start_char);
+                }
+                // We return the end char if the cursor is either on the start char
+                // or at some arbitrary position between start and end char.
+                return Some(end_char);
             }
-            // We return the end char if the cursor is either on the start char
-            // or at some arbitrary position between start and end char.
-            return Some(end_char);
+        }
+        // this node itselt wasn't a pair but maybe its siblings are
+
+        // check if we are *on* the pair (special cased so we don't look
+        // at the current node twice and to jump to the start on that case)
+        if let Some(open) = as_close_pair(doc, &node) {
+            if let Some(pair_start) = find_pair_end(doc, node.prev_sibling(), open, Backward) {
+                return Some(pair_start);
+            }
         }
 
-        if traverse_parents {
-            node = node.parent()?;
-        } else {
-            return None;
+        if !traverse_parents {
+            // check if we are *on* the opening pair (special cased here as
+            // an opptimization since we only care about bracket on the cursor
+            // here)
+            if let Some(close) = as_open_pair(doc, &node) {
+                if let Some(pair_end) = find_pair_end(doc, node.next_sibling(), close, Forward) {
+                    return Some(pair_end);
+                }
+            }
+            if node.is_named() {
+                return None;
+            }
         }
+
+        for close in
+            iter::successors(node.next_sibling(), |node| node.next_sibling()).take(MATCH_LIMIT)
+        {
+            let Some(open) = as_close_pair(doc, &close) else { continue; };
+            if find_pair_end(doc, Some(node), open, Backward).is_some() {
+                return doc.try_byte_to_char(close.start_byte()).ok();
+            }
+        }
+        node = node.parent()?;
     }
 }
 
@@ -159,6 +191,60 @@ fn surrounding_bytes(doc: &Rope, node: &Node) -> Option<(usize, usize)> {
     }
 
     Some((start_byte, end_byte))
+}
+
+/// Checks if `node` or its siblings (at most MATCH_LIMIT nodes) is the specified openiing char
+///
+/// # Returns
+///
+/// The position of the found node or `None` otherwise
+fn find_open_pair(doc: &Rope, node: Option<Node>, open: char) -> Option<usize> {
+    iter::successors(node, |node| node.prev_sibling())
+        .take(MATCH_LIMIT)
+        .find_map(|node| {
+            let (pos, c) = as_char(doc, &node)?;
+            (c == open).then_some(pos)
+        })
+}
+
+/// Tests if this node is a pair close char and returns the expected open char
+fn as_close_pair(doc: &Rope, node: &Node) -> Option<char> {
+    let close = as_char(doc, node)?.1;
+    PAIRS
+        .iter()
+        .find_map(|&(open, close_)| (close_ == close).then_some(open))
+}
+
+/// Checks if `node` or its siblings (at most MATCH_LIMIT nodes) is the specified closing char
+///
+/// # Returns
+///
+/// The position of the found node or `None` otherwise
+fn find_close_pair(doc: &Rope, node: Option<Node>, close: char) -> Option<usize> {
+    iter::successors(node, |node| node.next_sibling())
+        .take(MATCH_LIMIT)
+        .find_map(|node| {
+            let (pos, c) = as_char(doc, &node)?;
+            (end_char == c).then_some(pos)
+        })
+}
+
+/// Tests if this node is a pair close char and returns the expected open char
+fn as_open_pair(doc: &Rope, node: &Node) -> Option<char> {
+    let close = as_char(doc, node)?.1;
+    PAIRS
+        .iter()
+        .find_map(|&(open_, close)| (open_ == open).then_some(close))
+}
+
+/// Tests if this node is a pair opening and returns the expected close char
+fn as_char(doc: &Rope, node: &Node) -> Option<(usize, char)> {
+    // TODO: multi char/non ASCII pairs
+    if node.byte_range().len() != 1 {
+        return None;
+    }
+    let pos = doc.try_byte_to_char(node.start_byte()).ok()?;
+    Some((pos, doc.char(pos)))
 }
 
 #[cfg(test)]

--- a/helix-core/src/match_brackets.rs
+++ b/helix-core/src/match_brackets.rs
@@ -64,9 +64,12 @@ fn find_pair(syntax: &Syntax, doc: &Rope, pos: usize, traverse_parents: bool) ->
                 if end_byte == pos {
                     return Some(start_char);
                 }
+
                 // We return the end char if the cursor is either on the start char
                 // or at some arbitrary position between start and end char.
-                return Some(end_char);
+                if traverse_parents || start_byte == pos {
+                    return Some(end_char);
+                }
             }
         }
         // this node itselt wasn't a pair but maybe its siblings are

--- a/helix-core/src/match_brackets.rs
+++ b/helix-core/src/match_brackets.rs
@@ -1,8 +1,10 @@
 use std::iter;
 
+use ropey::RopeSlice;
 use tree_sitter::Node;
 
-use crate::{Rope, Syntax};
+use crate::movement::Direction::{self, Backward, Forward};
+use crate::Syntax;
 
 const MAX_PLAINTEXT_SCAN: usize = 10000;
 const MATCH_LIMIT: usize = 16;
@@ -27,7 +29,7 @@ const PAIRS: &[(char, char)] = &[
 ///
 /// If no matching bracket is found, `None` is returned.
 #[must_use]
-pub fn find_matching_bracket(syntax: &Syntax, doc: &Rope, pos: usize) -> Option<usize> {
+pub fn find_matching_bracket(syntax: &Syntax, doc: RopeSlice, pos: usize) -> Option<usize> {
     if pos >= doc.len_chars() || !is_valid_bracket(doc.char(pos)) {
         return None;
     }
@@ -45,13 +47,18 @@ pub fn find_matching_bracket(syntax: &Syntax, doc: &Rope, pos: usize) -> Option<
 //
 // If no surrounding scope is found, the function returns `None`.
 #[must_use]
-pub fn find_matching_bracket_fuzzy(syntax: &Syntax, doc: &Rope, pos: usize) -> Option<usize> {
+pub fn find_matching_bracket_fuzzy(syntax: &Syntax, doc: RopeSlice, pos: usize) -> Option<usize> {
     find_pair(syntax, doc, pos, true)
 }
 
-fn find_pair(syntax: &Syntax, doc: &Rope, pos: usize, traverse_parents: bool) -> Option<usize> {
+fn find_pair(
+    syntax: &Syntax,
+    doc: RopeSlice,
+    pos_: usize,
+    traverse_parents: bool,
+) -> Option<usize> {
     let tree = syntax.tree();
-    let pos = doc.char_to_byte(pos);
+    let pos = doc.char_to_byte(pos_);
 
     let mut node = tree.root_node().descendant_for_byte_range(pos, pos)?;
 
@@ -92,7 +99,7 @@ fn find_pair(syntax: &Syntax, doc: &Rope, pos: usize, traverse_parents: bool) ->
                 }
             }
             if node.is_named() {
-                return None;
+                break;
             }
         }
 
@@ -104,8 +111,16 @@ fn find_pair(syntax: &Syntax, doc: &Rope, pos: usize, traverse_parents: bool) ->
                 return doc.try_byte_to_char(close.start_byte()).ok();
             }
         }
-        node = node.parent()?;
+        let Some(parent) = node.parent() else { break; };
+        node = parent;
     }
+    let node = tree.root_node().named_descendant_for_byte_range(pos, pos)?;
+    if node.child_count() != 0 {
+        return None;
+    }
+    let node_start = doc.byte_to_char(node.start_byte());
+    find_matching_bracket_plaintext(doc.byte_slice(node.byte_range()), pos_ - node_start)
+        .map(|pos| pos + node_start)
 }
 
 /// Returns the position of the matching bracket under cursor.
@@ -120,10 +135,7 @@ fn find_pair(syntax: &Syntax, doc: &Rope, pos: usize, traverse_parents: bool) ->
 ///
 /// If no matching bracket is found, `None` is returned.
 #[must_use]
-pub fn find_matching_bracket_current_line_plaintext(
-    doc: &Rope,
-    cursor_pos: usize,
-) -> Option<usize> {
+pub fn find_matching_bracket_plaintext(doc: RopeSlice, cursor_pos: usize) -> Option<usize> {
     // Don't do anything when the cursor is not on top of a bracket.
     let bracket = doc.char(cursor_pos);
     if !is_valid_bracket(bracket) {
@@ -179,11 +191,11 @@ fn is_forward_bracket(c: char) -> bool {
     PAIRS.iter().any(|(l, _)| *l == c)
 }
 
-fn is_valid_pair(doc: &Rope, start_char: usize, end_char: usize) -> bool {
+fn is_valid_pair(doc: RopeSlice, start_char: usize, end_char: usize) -> bool {
     PAIRS.contains(&(doc.char(start_char), doc.char(end_char)))
 }
 
-fn surrounding_bytes(doc: &Rope, node: &Node) -> Option<(usize, usize)> {
+fn surrounding_bytes(doc: RopeSlice, node: &Node) -> Option<(usize, usize)> {
     let len = doc.len_bytes();
 
     let start_byte = node.start_byte();
@@ -196,22 +208,8 @@ fn surrounding_bytes(doc: &Rope, node: &Node) -> Option<(usize, usize)> {
     Some((start_byte, end_byte))
 }
 
-/// Checks if `node` or its siblings (at most MATCH_LIMIT nodes) is the specified openiing char
-///
-/// # Returns
-///
-/// The position of the found node or `None` otherwise
-fn find_open_pair(doc: &Rope, node: Option<Node>, open: char) -> Option<usize> {
-    iter::successors(node, |node| node.prev_sibling())
-        .take(MATCH_LIMIT)
-        .find_map(|node| {
-            let (pos, c) = as_char(doc, &node)?;
-            (c == open).then_some(pos)
-        })
-}
-
 /// Tests if this node is a pair close char and returns the expected open char
-fn as_close_pair(doc: &Rope, node: &Node) -> Option<char> {
+fn as_close_pair(doc: RopeSlice, node: &Node) -> Option<char> {
     let close = as_char(doc, node)?.1;
     PAIRS
         .iter()
@@ -223,8 +221,17 @@ fn as_close_pair(doc: &Rope, node: &Node) -> Option<char> {
 /// # Returns
 ///
 /// The position of the found node or `None` otherwise
-fn find_close_pair(doc: &Rope, node: Option<Node>, close: char) -> Option<usize> {
-    iter::successors(node, |node| node.next_sibling())
+fn find_pair_end(
+    doc: RopeSlice,
+    node: Option<Node>,
+    end_char: char,
+    direction: Direction,
+) -> Option<usize> {
+    let advance = match direction {
+        Forward => Node::next_sibling,
+        Backward => Node::prev_sibling,
+    };
+    iter::successors(node, advance)
         .take(MATCH_LIMIT)
         .find_map(|node| {
             let (pos, c) = as_char(doc, &node)?;
@@ -233,15 +240,15 @@ fn find_close_pair(doc: &Rope, node: Option<Node>, close: char) -> Option<usize>
 }
 
 /// Tests if this node is a pair close char and returns the expected open char
-fn as_open_pair(doc: &Rope, node: &Node) -> Option<char> {
-    let close = as_char(doc, node)?.1;
+fn as_open_pair(doc: RopeSlice, node: &Node) -> Option<char> {
+    let open = as_char(doc, node)?.1;
     PAIRS
         .iter()
         .find_map(|&(open_, close)| (open_ == open).then_some(close))
 }
 
-/// Tests if this node is a pair opening and returns the expected close char
-fn as_char(doc: &Rope, node: &Node) -> Option<(usize, char)> {
+/// If node is a single char return it (and its char position)
+fn as_char(doc: RopeSlice, node: &Node) -> Option<(usize, char)> {
     // TODO: multi char/non ASCII pairs
     if node.byte_range().len() != 1 {
         return None;
@@ -257,11 +264,11 @@ mod tests {
     #[test]
     fn test_find_matching_bracket_current_line_plaintext() {
         let assert = |input: &str, pos, expected| {
-            let input = &Rope::from(input);
-            let actual = find_matching_bracket_current_line_plaintext(input, pos);
+            let input = RopeSlice::from(input);
+            let actual = find_matching_bracket_plaintext(input, pos);
             assert_eq!(expected, actual.unwrap());
 
-            let actual = find_matching_bracket_current_line_plaintext(input, expected);
+            let actual = find_matching_bracket_plaintext(input, expected);
             assert_eq!(pos, actual.unwrap(), "expected symmetrical behaviour");
         };
 

--- a/helix-term/src/commands.rs
+++ b/helix-term/src/commands.rs
@@ -4535,8 +4535,8 @@ fn match_brackets(cx: &mut Context) {
     let selection = doc.selection(view.id).clone().transform(|range| {
         let pos = range.cursor(text_slice);
         if let Some(matched_pos) = doc.syntax().map_or_else(
-            || match_brackets::find_matching_bracket_current_line_plaintext(text, pos),
-            |syntax| match_brackets::find_matching_bracket_fuzzy(syntax, text, pos),
+            || match_brackets::find_matching_bracket_plaintext(text.slice(..), pos),
+            |syntax| match_brackets::find_matching_bracket_fuzzy(syntax, text.slice(..), pos),
         ) {
             range.put_cursor(text_slice, matched_pos, is_select)
         } else {

--- a/helix-term/src/ui/editor.rs
+++ b/helix-term/src/ui/editor.rs
@@ -501,7 +501,9 @@ impl EditorView {
             use helix_core::match_brackets;
             let pos = doc.selection(view.id).primary().cursor(text);
 
-            if let Some(pos) = match_brackets::find_matching_bracket(syntax, doc.text(), pos) {
+            if let Some(pos) =
+                match_brackets::find_matching_bracket(syntax, doc.text().slice(..), pos)
+            {
                 // ensure col is on screen
                 if let Some(highlight) = theme.find_scope_index_exact("ui.cursor.match") {
                     return vec![(highlight, pos..pos + 1)];

